Implement gas blender - Mobile

fnightfox opened the pull request:
### Describe the pull request:
<!-- Replace [ ] with [x] to select options. -->
- [ ] Bug fix
- [ ] Functional change
- [X] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Pull request long description:
This pull request implements a gas blender and ties this functionality into the mobile app, technical diving->gas calculator

The algorithm I implemented is not one that I found online or read anywhere. I am looking to write it up in a formal way and (ideally) publish it somewhere for further review, but so far it seems accurate and complete. The basics involve the following:

1. Convert EVERYTHING into volumes. We should deal with pressures as little as possible, and when we do try to include the real gas law vs the ideal gas law. To this end, you will see a copy of the tables used elsewhere in subsurface in the core/gasblender cpp file. Liner approximation is used for values between those presented, and should be accurate enough for this application.

2. Take all gas sources and split them into two types: unlimited (called "GasSources") and limited gasses
3. The limited gas sources are used to create "Blend" objects. These objects are order dependent, and need to have every combination possible, including draining the target cylinder to ATM. This causes a factorial explosion such that only 3 limited volume sources cause ~16 distinct blends. 
4. These are grouped by order used in the steps section of the struct, and then every combination of 3 blends is used to create a triangle where the x,y coordinate is the neededO2, neededHe for the REMAINING gas to fill the target cylinder to full with our desired mix.

5. All of the GasSource objects are used to form as many distinct triangles as possible. These represent the POSSIBLE blends of unlimited gas we have access to.

6. We check for any overlap between the two sets of triangles. This is the computationally expensive step.

7. Each overlapping "point of interest" (where the triangles intersect, any point of one triangle contained in another, etc..)  is filled out as a blend. These are saved.

8. All of those blends are checked for cost. The lowest cost blend is returned.

I did a decent amount of checking and validation to ensure this algorithm is sound, but if there were any questions please raise them in the comments.

**Some items of note:**
1. You must have at least one unlimited source
2. Simple Blending makes all sources unlimited
3. Source order for unlimited sources leans towards the order they are on your source list
4. PRESSURE calculations are only done when needed, and the target pressure is only used at the beginning. The pressure calculated at the end is not checked against the target and should be used to assure yourself of the accuracy of the calculations. I noted a ~9PSI difference when targeting 3000PSI. This is acceptable and represents a .3% calculation error throughout the calculations. 
5. The final mix should be within .1% of the target.
6. I am interested in blending by weight, as it does not depend on the temperature of the gas or accuracy of the pressure gauge, so that is why I put that value first (and it is more accurate than pressure is anyway).

**PLEASE DO NOT MERGE UNTIL THIS HAS BEEN TESTED, DEBUG CODE REMOVED, AND (IDEALLY) SOME DOCUMENTATION HAS BEEN CREATED**

Items I know need to be done:
1. Convert spaces to tabs
2. Remove the commented out code using the convexHull. This method greatly reduces the number of triangles, but does not work as a minimum cost method. I have left it in because I do not want to delete it until others comment.

### Changes made:
1. Add GasBlender cpp core code
2. Add unit tests for the gas blender cpp code
3. Add Gas Calculator mobile front end codeub that this PR fixes an issue. -->

### Additional information:
See long description for notes and algorithm

### Documentation change:
None, yet

### Mentions:

fnightfox committed:
Initial completed Gas Blender
Signed-off-by: liz <liz.g.dunlap@gmail.com>

mikeller commented:
Wow, this is certainly a **very** comprehensive blender! Good work.

I think @atdotde is best positioned to comment on the mathematical approach taken.

From a user perspective this is already quite good to use, just a couple of rough edges:
- use a better default than 3000 if the pressure is set to bar (or persist the user's preferences for these settings);
- since the ordering of gases in the list dictates the order of filling for the simple blender, allow users to reorder rows without having to delete / re-create.

Some ideas for possible improvements:
- in the full blender, allow a limited gas to be boosted - still only a limited volume available, but not limited by blending only to equilibrium;
- in the full blender, allow the user to create 'blend plans' - when a desired mix is accepted by the user, store it for later retrieval, and update the gas list with the remaining gas after the blend was done, so they can plan the next blend with what is left in the supply tanks.

atdotde commented:
So far I only had a brief look, but this looks far too complicated to me. And you are reinventing the whee in far too many places. Subsurface already has all required helper functions to deal with gas calculations we we took great care to get those right (in particular for dealing with real gases). Please use those! 

Also I think the approach to express everything in volumes is wrong. The calculation does not depend on intensive variables like pressure, then also the formulas get simpler and only in the very end, you multiply by the actual volume of the cylinder.

I also don't see where you need any iterations over filling steps. The amounts of gas are determined by a system of linear equations:

You start from a given mixture at a given pressure and have a target mixture and pressure. You can convert both gases to say 1bar to find the amount (if you like in linters) of each of the partial gases that are required to be added. Since you need to add a specific number of litres of O2, N2 and He, you need three linearly independent sources of gas to add (if you are blending nitrox, everything reduces to 2). But those sources in general are not pure but are certain pictures. Finding the amount of each source is done by solving a system of three linear equations and three unknown, You can use closed formulas for that, even implementing a matrix inverse is overkill. Once you know how much of each source you need at 1bar you can (using again the compressibility factor) convert that to the different intermediate pressures. And at that point you could check that your limited sources actually contain enough gas to allow for that plan.

All this  is implemented in lines 123--214 of https://github.com/atdotde/realblender/blob/main/blender.pl , It doesn't have to be more complicated than that.

Or am I missing something?

fnightfox commented:
The complications mostly come from the need to limit the amount of gas. In the solution you stated the viability of a mixture is only checked at the end, and even then there is no way to validate if it is the cheapest possible blend.

I will look through the core code again for helper functions to see what's been implemented and what I can pull into this, but I do believe the method you detailed is not comprehensive enough for the use case I was imagining, where a limited gas can be used up to the limit and then the blend is continued from there, finding the cheapest option.

dirkhh commented:
a gas blending feature that allows (based on your description) unlimited number of input gases with additional input values (the available pressure, the availability of a booster, and, I assume, given your desire to match a realistic use case, the performance characteristics of said booster) is a very specific, very narrow, and by necessity very complex UI.

I encourage you to build an Android or iOS app with that exact feature set.

I do not agree that this is a feature that should be added to Subsurface.

mikeller commented:
> I would say that this is a cool niche application for the nine people on earth who want to bother with it to save 6 bucks. And I think that this complexity (especially the UI side of it) is not useful or appreciated (at least from my perspective) for Subsurface as that is far, far, FAR outside the target audience.

Full disclosure: I am another one of the nine. 😉

But joke aside, I think most technical divers doing their own trimix blending fall into the 'limited gases' category. And in many parts of the world this is the only option to do trimix diving, as there simply aren't any dive shops that are set up and willing to sell trimix fills.

And I think having the option to select between the 'simple blender' UI which will be good for most Nitrox applications, and the 'full blender', which will be preferable for most Trimix blending is a good way to solve the UI complexity. Possibly it could be turned into a preference, assuming that most users will want to use either one or the other.

fnightfox commented:
Attached to this commend is a quick write up I did on this methodology. I am still tracking down a bug in handling the hyperbolic problem of limited gas sources (this needs to be handled in two places, I have solved one but need to figure out the other. It really only comes up when limited sources that are very large are used, ones that could be handled as unlimited in the current implementation and give workable results, but I want to get it nailed down completely before merging, if this is the route we decide to go down).

I am posting this as a way to get discussion on this method vs others. I am happy to recode the entire thing if another method makes more sense, but I don't want to just because this is a new way of approaching the problem, as I feel this solves some of the tricky aspects of blending (like limited sources, ordering the sources, and finding minimum cost).

Still TODO:
1. fix UI issues from the move to QT6
2. Check existing gas functions for possibility of reuse
 - dealing with absolute vs gauge pressure
 - dealing with gas compressibility
3. Track down method of handling the hyperbolic issue for trimix.  
 - UPDATE: I think I solved this by fixing the volumes of limited gasses to clamp, in order they are added, to max at the target value for the cylinder being filled. Not pushing this yet, until I get the UI fixes in.

[blending-paper-1.pdf](https://github.com/user-attachments/files/26576244/blending-paper-1.pdf)

fnightfox commented:
Ok, major changes in the most recent version. I reworked the entire system to use linear programming with a simplex matrix and solver., which is more traditional for solving this problem So please ignore the document I sent earlier, as it no longer represents the code in this PR. I have updated the blender to use the compression factor/real gas constants generated by the rest of the app, but I do not think there is a real opportunity for reuse for the sections dealing with the absolute pressure/gauge pressure conversions, since that existing code is mainly used with dive planning or dive objects and one does not exist in the blending context.

I have added a set of unit tests as well, and updated the UI for QT6. 

At this point, I am content with this merging if people want/like this implementation, and if the UI is accepted I can work on the documentation (at lease as much as I can do, I have never worked on the documentation in this project before).

Thank you all,
--Liz

mikeller reviewed (commented):
This looks by and large ok to me - I haven't tested it yet.

@atdotde, @bstoeger can you please have a look?

fnightfox commented:
I just noticed weird behavior in other parts of the app. I want a chance to test if it's my fault/this PR or unrelated. The SAC rate on logged dives is off now.

fnightfox commented:
@mikeller I have checked and the bug IS present in 5660 nightly release. It appears that the SAC on mobile that is displayed is either only for one tank or is 1/[# of tanks]. I dive sidemount and my SAC listed is almost exactly 1/2 of what I expect it to be. This appears to only happen when there are more than one tank present. I will look into it tomorrow if I have time, otherwise I will submit a bug report. I have no issues with this PR continuing to be evaluated/merged or declined.

--Liz

fnightfox commented:
ok. I have tracked down the issue I was seeing, and it probably comes down to a setting on my perdix. My second tank was being imported as "not used" vs OC Gas, and the recent merge made it such that the tank was not being counted. As a result, this would probably be fixed with a dropdown on the edit dive screen for "use" for the cylinder, but it is not in fact a bug.
